plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -127,3 +154,235 @@ counted out already so we should not immediately encounter those in `VBody`.
+          isAcceptable <- acceptable subst
+          -- if the size and cost are not acceptable, don't inline
+          if not isAcceptable then pure v
+          else do -- if the size and cost are acceptable, then check if it's fully applied
+            -- it's fully applied if its annotation (`a`) is in the list of fully applied called
+            -- locations. See note [Identifying fully applied call sites].
+            pure $ if a `elem` calledLocation info then subst else v
+    considerInline _notVar = -- this should not happen
+      Prelude.error  "considerInline: should be a variable."
+
+{- Note [Identifying fully applied call sites]
+To identify call sites that are "fully applied" (see note [Inlining of fully applied functions] for
+our definition of fully applied), we examine the `body` of the let term with `countApp` (a pure
+function).
+We track all term or type applications, and add them to a map if they are associated with a
+variable. This map contains all variable call sites, each call site is uniquely identified by an
+annotation (the annotation therefore has to be unique and constant).
+We then compare this with the variable's lambda abstraction order. If any call sites application
+order is the same as the lambda abstraction order, we add the annotations of these call sites into
+the `calledLocation` field of `Utils.CalledVarInfo`, which lists the annotations of a variable's
+fully applied called locations. We only inline a variable if its annotation is in this list.
+-}
+
+-- | Run the inliner on a single non-recursive let binding.
+processSingleBinding
+    :: forall tyname name uni fun a. (InliningConstraints tyname name uni fun, Eq a, Ord name)
+    => Term tyname name uni fun a -- ^ The body of the let binding.
+    -> Binding tyname name uni fun a -- ^ The binding.
+    -> InlineM tyname name uni fun a (Binding tyname name uni fun a)
+processSingleBinding body = \case
+    -- when the let binding is a function type, we add it to the `CalledVarEnv` and
+    -- consider whether we want to inline at the call site.
+    TermBind a s v@(VarDecl _ n (TyFun _ _tyArg _tyBody)) rhs -> do
+        let
+          -- track the term and type lambda abstraction order of the function
+          varLamOrder = countLam rhs
+          -- examine the `body` of the `Let` term and track all term/type applications.
+          appSites = countApp body
+          -- list of all call sites of this variable
+          listOfCallSites = Map.lookup n appSites
+        case listOfCallSites of
+          Nothing ->
+            -- we don't remove the binding because we decide *at the call site* whether we want to
+            -- inline, and it may be called more than once
+            pure $ TermBind a s v rhs
+          Just list -> do
+            let
+              isEqAppOrder :: ApplicationOrder a -> Bool
+              isEqAppOrder appOrder = applicationOrder appOrder == varLamOrder
```

Comment:
I don't think `appOrder` and `varLamOrder` has to be equal. `varLamOrder` is allowed to be a prefix of `appOrder`.

Reply:
Over-application of a function/variable should not type check and thus should not happen? (As mentioned in the note)

Reply:
An example of over-application:

```haskell
let id = \y -> y
     f = \x -> id
 in f x y
```

`f`'s arity is 1, but is applied to 2 arguments.